napi_basics.md

```diff
@@ -0,0 +1,139 @@
+# N-API - Basics
+
+N-API is an API exposed by Node to developers that allow them to author Native 
+add-ons in an ABI stable manner. This means that these add-ons can work 
+against future versions of Node without recompilation, and developers get the 
+additional benefit that their add-ons can work independent of the underlying 
+JavaScript Virtual Machine being used by Node.
+
+APIs exposed in the N-API surface are generally used to create and manipulate
+JavaScript objects. Concepts and operations generally map to ideas specified 
+in the ECMA262 Language Specification. The APIs generally have the following 
+properties:
+- All N-API APIs generally return a status code of type `napi_status`. This
+  status can be used to determine whether the API call succeeded or failed.
+- The actual value being returned by the API is returned in an out parameter.
+- All JavaScript values are abstracted behind an opaque type named 
+  `napi_value`.
+- In case of an error status code, additional information can be obtained 
+  using `napi_get_last_error_info`. More information can be found in the error
+  handling section (<PLACEHOLDER: link to error handling doc>)
+
+## N-API General types
+
+N-API exposes the following fundamental datatypes as abstractions that are 
+consumed by the various APIs. These APIs should be treated as opaque types 
+that can be introspected only with other N-API APIs.
+
+### *napi_status*
+Integral status code indicating the success or failure of a N-API API call.
+Currently, the following status codes are supported.
+```C
+  napi_ok,
+  napi_invalid_arg,
+  napi_object_expected,
+  napi_string_expected,
+  napi_function_expected,
+  napi_number_expected,
+  napi_boolean_expected,
+  napi_generic_failure,
+  napi_pending_exception,
+  napi_status_last
+```
+If additional information is required upon an API returning a failed status, 
+it can be obtained by calling `napi_get_last_error_info`.
+
+### *napi_extended_error_info*
+#### Definition
+```C
+struct napi_extended_error_info {
+  const char* error_message;
+  void* engine_reserved;
+  uint32_t engine_error_code;
+  napi_status error_code;
+};
+```
+
+#### Members
+- error_message: UTF8-encoded string containing a VM-neutral description of 
+  the error
+- engine_reserved: Reserved for VM-specific error details. This is currently
+  not implemented for any VM.
+- engine_error_code: VM-specific error code
```

Comment:
This is also currently unimplemented, IIRC. If so, this should also have "This is currently not implemented for any VM."